xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/suggest/suggestPages.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+require.config({
+  paths: {
+    'xwiki-selectize': "$xwiki.getSkinFile('uicomponents/suggest/xwiki.selectize.js', true)" +
+      "?v=$escapetool.url($xwiki.version)"
+  }
+});
+
+define('xwiki-suggestPages', ['jquery', 'xwiki-selectize'], function($) {
+  'use strict';
+
+  var pageIcon = $jsontool.serialize($services.icon.getMetaData('page_white'));
+
+  var getSelectizeOptions = function(select) {
+    return {
+      create: true,
+      load: function(text, callback) {
+        loadPages(text).done(function(data) {
+          var pages = [];
+          data.searchResults.forEach(function (element) {
+            var label = element.title;
+            var queryString = "";
+            // Separate spaces with " / " and unescape characters. E.g. A\.B.C\\D => A.B / C\D
+            var hint = element.space.replace(/([^\\])\./g, "$1 / ").replace(/\\(.)/g, "$1");
+            if (element.language) {
+              queryString = "language=" + element.language;
+              hint += " (" + element.language + ")";
+            }
+            var url = new XWiki.Document(XWiki.Model.resolve(element.id, XWiki.EntityType.DOCUMENT))
+              .getURL("view", queryString);
+            pages.push({
+              'value': element.pageFullName,
+              'label': label,
+              'icon': pageIcon,
+              'url': url,
+              'hint': hint
+            });
+          });
+          callback(pages);
+        }).fail(callback);
+      }
+    }
+  };
+
+  var loadPages = function(text) {
+    // We need to escape backslashes for Solr.
+    text = text.replace(/\\/g, "\\\\");
+    var response = $.getJSON("${request.contextPath}/rest/wikis/query/", {
+      'q': "title:*" + text + "* or fullname:*" + text + "*",
```

Comment:
Computing the Solr query from the user input can lead to unexpected results. For instance if the text is "one two" then your query is ``title:*one two*`` which means:

* match the titles that **end** with "one"
* match **any field** that starts with "two"

You could use [grouping](https://lucene.apache.org/solr/guide/6_6/the-standard-query-parser.html#TheStandardQueryParser-GroupingClauseswithinaField) but you'd get ``title:(*one two*)`` which means:

* match the titles that **end** with "one"
* match the titles that **start** with "two"

And this can go on. The bottom line is: avoid computing the Solr query from the user input. The safest is to pass the user input as Solr query  without any modification and to use the query parameters to restrict / tune the search. For instance instead of doing:

    q=title:text or fullname:text

you could write:

    q=text
    qf=title fullname

Which means: look for "text" in the specified query fields (title and fullname). I don't know if the ``/rest/wikis/query/`` REST resource supports specifying the query fields parameter, but it probably should.

Now, even with this you still have the problem with the wildcards. Note that the live search (search suggest) from the top right corder of the page doesn't use wildcards and neither the tree finder, but there have been complains about this. The real fix for this is to index more information, e.g. using [NGrams](http://blog.florian-hopf.de/2014/03/prefix-and-suffix-matches-in-solr.html). In other words, with Solr, if there's a search problem you don't fix the query, you fix the index (most of the time).

Reply:
> The real fix for this is to index more information, e.g. using NGrams. In other words, with Solr, if there's a search problem you don't fix the query, you fix the index (most of the time).

Yes I thought about that but can we change the Solr configuration?

Reply:
We can of course but we need to be aware of the consequences. Indexing prefixes/suffixes can grow the size of the index and can increase the index time. The search time shouldn't be affected normally. In any case, this is something to discuss on the mailing list and we need to see if it fits in the roadmap.